internal/deploy/application/remote.go

```diff
@@ -0,0 +1,288 @@
+		exe.catalogByName[listCatalog.CatalogItems[i].Name] = &listCatalog.CatalogItems[i]
+	}
+	return
+}
+
+func (exe *remoteExecutor) validate() (err error) {
+	// Validate routes
+	for _, route := range exe.opt.Routes {
+		if _, foundFrom := exe.microserviceByName[route.From]; !foundFrom {
+			return util.NewNotFoundError(fmt.Sprintf("Could not find origin microservice for the route %v", route))
+		}
+		if _, foundTo := exe.microserviceByName[route.To]; !foundTo {
+			return util.NewNotFoundError(fmt.Sprintf("Could not find destination microservice for the route %v", route))
+		}
+	}
+
+	// Validate microservice
+	for _, msvc := range exe.opt.Microservices {
+		if _, foundAgent := exe.agentsByName[msvc.Agent.Name]; !foundAgent {
+			return util.NewNotFoundError(fmt.Sprintf("Could not find agent: %s", msvc.Agent.Name))
+		}
+		if _, foundCatalogItem := exe.catalogByID[msvc.Images.CatalogID]; msvc.Images.CatalogID > 0 && !foundCatalogItem {
+			return util.NewNotFoundError(fmt.Sprintf("Could not find catalog item: %d", msvc.Images.CatalogID))
+		}
+	}
+
+	// TODO: Check if application alredy exists
+	return nil
+}
+
+func (exe *remoteExecutor) deploy() (err error) {
+	defer util.SpinStop()
+
+	// Create flow
+	util.SpinStart("Creating flow")
+	flow, err := exe.client.CreateFlow(exe.opt.Name, fmt.Sprintf("Flow for application: %s", exe.opt.Name))
+	if err != nil {
+		return
+	}
+
+	// Create microservices
+	for _, msvc := range exe.opt.Microservices {
+		util.SpinStart(fmt.Sprintf("Deploying microservice %s", msvc.Name))
+
+		// TODO: Configure agent
+		agent, _ := exe.agentsByName[msvc.Agent.Name]
+		_, err = exe.client.UpdateAgent(&client.AgentUpdateRequest{
+			UUID: agent.UUID,
+			AgentConfiguration: client.AgentConfiguration{
+				DockerURL:                 msvc.Agent.Config.DockerURL,
+				DiskLimit:                 msvc.Agent.Config.DiskLimit,
+				DiskDirectory:             msvc.Agent.Config.DiskDirectory,
+				MemoryLimit:               msvc.Agent.Config.MemoryLimit,
+				CPULimit:                  msvc.Agent.Config.CPULimit,
+				LogLimit:                  msvc.Agent.Config.LogLimit,
+				LogDirectory:              msvc.Agent.Config.LogDirectory,
+				LogFileCount:              msvc.Agent.Config.LogFileCount,
+				StatusFrequency:           msvc.Agent.Config.StatusFrequency,
+				ChangeFrequency:           msvc.Agent.Config.ChangeFrequency,
+				DeviceScanFrequency:       msvc.Agent.Config.DeviceScanFrequency,
+				BluetoothEnabled:          msvc.Agent.Config.BluetoothEnabled,
+				WatchdogEnabled:           msvc.Agent.Config.WatchdogEnabled,
+				AbstractedHardwareEnabled: msvc.Agent.Config.AbstractedHardwareEnabled,
+			},
+		})
+		if err != nil {
+			return
+		}
+		// CatalogItem
+		var catalogItem *client.CatalogItemInfo
+		catalogImages := []client.CatalogImage{
+			{ContainerImage: msvc.Images.X86, AgentTypeID: 1},
+			{ContainerImage: msvc.Images.ARM, AgentTypeID: 2},
+		}
+		if msvc.Images.CatalogID == 0 {
```

Comment:
This function gets really hard to follow at this point because of all the indentation - can we break it up into more functions?

Reply:
I understand the bulk of the lines is just type decl/definitions

Reply:
Done